File: java/v1/src/main/java/com/api/v1/services/payment/PaymentRegistrationServiceImpl.java
package com.api.v1.services.payment;

import com.api.v1.annotations.car.VIN;
import com.api.v1.annotations.creditcard.CardNumber;
import com.api.v1.annotations.customer.SSN;
import com.api.v1.builders.payment.PaymentBuilder;
import com.api.v1.domain.car.Car;
import com.api.v1.domain.creditcard.CreditCard;
import com.api.v1.domain.customer.Customer;
import com.api.v1.domain.payment.Payment;
import com.api.v1.domain.payment.PaymentRepository;
import com.api.v1.dtos.payment.PaymentResponseDto;
import com.api.v1.mappers.payment.PaymentResponseMapper;
import com.api.v1.utils.car.CarFinderUtil;
import com.api.v1.utils.creditcard.CreditCardFinderUtil;
import com.api.v1.utils.customer.CustomerFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class PaymentRegistrationServiceImpl implements PaymentRegistrationService {

    @Autowired
    private PaymentRepository repository;

    @Autowired
    private CustomerFinderUtil customerFinderUtil;

    @Autowired
    private CarFinderUtil carFinderUtil;

    @Autowired
    private CreditCardFinderUtil creditCardFinderUtil;

    @Override
    public Mono<PaymentResponseDto> register(@SSN String ssn, @VIN String vin, @CardNumber String cardNumber) {
        Mono<Customer> customerMono = customerFinderUtil.find(ssn);
        Mono<Car> carMono = carFinderUtil.find(vin);
        Mono<CreditCard> creditCardMono = creditCardFinderUtil.find(cardNumber);
        return Mono.zip(customerMono, carMono, creditCardMono)
                .flatMap(tuple -> {
                    Customer customer = tuple.getT1();
                    Car car = tuple.getT2();
                    CreditCard creditCard = tuple.getT3();
                    Payment payment = PaymentBuilder
                            .create()
                            .withCustomer(customer)
                            .withCar(car)
                            .withCreditCard(creditCard)
                            .build();
                    return repository.save(payment)
                            .flatMap(savedPayment -> Mono.just(
                                    new PaymentResponseMapper(savedPayment, creditCard).map()));
                });
    }

}

File: v2/src/main/kotlin/com/api/v2/payment/services/PaymentRegistrationServiceImpl.kt
package com.api.v2.payment.services

import com.api.v2.car.utils.CarFinderUtil
import com.api.v2.creditcard.utils.CreditCardFinderUtil
import com.api.v2.customer.utils.CustomerFinderUtil
import com.api.v2.payment.domain.Payment
import com.api.v2.payment.domain.PaymentRepository
import com.api.v2.payment.dtos.PaymentResponseDto
import com.api.v2.payment.mappers.PaymentResponseMapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class PaymentRegistrationServiceImpl: PaymentRegistrationService {

    @Autowired
    lateinit var customerFinderUtil: CustomerFinderUtil

    @Autowired
    lateinit var carFinderUtil: CarFinderUtil

    @Autowired
    lateinit var creditCardFinderUtil: CreditCardFinderUtil

    @Autowired
    lateinit var paymentRepository: PaymentRepository

    override suspend fun register(ssn: String, vin: String, cardNumber: String): PaymentResponseDto {
        return withContext(Dispatchers.IO) {
            val customer = customerFinderUtil.find("123456789")
            val car = carFinderUtil.find("1234567890123456")
            val creditCard = creditCardFinderUtil.find("1234567890123456")
            val payment = Payment(customer, car, creditCard)
            val savedPayment =paymentRepository.save(payment)
            PaymentResponseMapper.map(savedPayment, customer, car, creditCard)
        }
    }

}